tools/packages/wesl/src/WeslDevice.ts

```diff
@@ -0,0 +1,159 @@
+import { ExtendedGPUValidationError } from "./LinkedWesl";
+
+type ErrorScope = {
+  filter: GPUErrorFilter;
+  errors: Promise<GPUError | null>[];
+};
+
+/**
+ * A {@link GPUDevice} with extensions for WESL. Created with {@link makeWeslDevice}.
+ */
+export interface WeslDevice extends GPUDevice {
+  /**
+   * Attaches an error to the current error scope (created by {@link GPUDevice.pushErrorScope}).
+   * If there is no error scope, it reports the error as a `'uncapturederror'`
+   */
+  injectError(type: GPUErrorFilter, error: Promise<GPUError | null>): void;
+}
+
+/**
+ * Mutates a {@link GPUDevice} for usage with WESL. Does not impact your existing code, wherever a {@link GPUDevice} can be used, a {@link WeslDevice} is also valid.
+ *
+ * WESL uses this to display errors pointing at the WESL source instead of pointing at generated code.
+ */
+export function makeWeslDevice(device: GPUDevice): WeslDevice {
```

Comment:
This keeps track of the `pushErrorScope` and `popErrorScope`s. That way, we get to inject our own error into them.